Pagination Component

langdonx opened the pull request:
Take a look and let me know what you think.  Some things of note:
 - It's not as customizable as UI Bootstrap's implementation, but I don't know that the extra options they provide are a good thing, as they have the potential to limit usability.
 - There's no 2-way binding at all (yay), but I did have to pull in `$scope` to `$watch` the bindings for changes and regenerate which pages are displayed. I don't know of a way around this.
 - The way I chose to implement `visible-page-buffer` might be controversial, but I strongly think that it provides the best usability possible.  It always bothered me that UI Bootstrap's implementation jumped around under my mouse as I clicked through pages.  I think clicking a moving target is a UX anti-pattern.
 - The logic in `buildPages()` is fuzzy. The algorithm could certainly be improved--it was adapted from PHP I wrote some 10+ years ago, but it's solid and handles all scenarios I've thrown at it in the past as well as some tests I had running earlier using `ng-repeat="items in $ctrl.generateRange(100)"` and variations of other attributes.
 - There are some comments in `src/library/pagination/pagination.scss`.  I left the CSS there because I think it helps usability, but I didn't like hard-coding those color values.
 - I had to adjust the columns on the demo page to account for the larger footprint of `ngbs-pagination`.  See `src/docs/utils/demo/demo.html`.
 - VS Code insisted on formatting my HTML with 4 spaces. I can fix if that's an issue.

langdonx committed:
stubbed out pagination

langdonx committed:
implemented pagination

langdonx commented:
This resolves #28.

IdanCo commented on src/docs/demos/pagination/pagination-demo.html:
```diff
@@ -0,0 +1,57 @@
+    <ngbs-pagination items-per-page="10" total-items="100" on-page-change="currentPageBasic10 = currentPage"></ngbs-pagination>
+</div>
+
+<div class="mb-4">
```
Love the demos, very comprehensive, but i don't think we need so many. sizes or disabled state are quit self-explanatory from the MD, no?

langdonx replied:
Probably not.  Half of them were for testing.  After I made the PR I thought it would be much better to have 1 example with other inputs to let the user adjust all the attributes.  The way the demo page is laid out though, it's very easy for the pagination component to run into the documentation on the right.

Is it OK if that happens if they put in crazy values, or can we somehow avoid it?  z-index maybe, and let it just float over the documentation?

IdanCo replied:
sounds good to me - crazy values lead to crazy behaviour, nothing wrong with that!

but if you want feel free to leave in just a hardcoded example, you made it very easy to understand

IdanCo commented on src/docs/demos/pagination/pagination-demo.html:
```diff
@@ -0,0 +1,57 @@
+            width: 80px;
+        }
+    </style>
+    <ngbs-pagination class="custom-pagination-1"
```
Also - awesome stuff, but not sure we need so many

IdanCo commented on src/docs/demos/pagination/pagination-demo.md:
```diff
@@ -0,0 +1,15 @@
+## ngbs-pagination
+
```
Great explanations!

IdanCo commented on src/library/pagination/pagination.component.js:
```diff
@@ -0,0 +1,170 @@
+    this.itemsPerPage = 0; // binding
+    this.size = ''; // binding
+    this.totalItems = 0; // binding
+    this.visiblePageBuffer = 0; // binding
```
why do we need these initializers? they're already injected through the binding, no?

langdonx replied:
It's a habit, really.  I will yank them out.

It's mainly for readability and maintenance purposes.  I always like that a class can stand on its own without needing to reference any documentation, or in this case, the component definition.  Since JavaScript isn't typed, these are simply hints to show the developer what values will end up in those properties.  I've often thought about using `Object.seal` in the constructor as well, to limit side effects, but haven't gone that far.  Sometimes it's nice to be able to set a property on an object you don't know... that's sort of the joy of JavaScript. ;)

IdanCo replied:
mmm... interesting point! never thought about it like that. I like independence...

I'm not sure what do, feel free to leave it in for now, we can always plunk it out later.

langdonx replied:
I'll yank it unless we decide it's valuable to have across the board.  I value consistency almost above all. ;) It is sort of a pain to manage anyway and the bindings are right there at the bottom.

IdanCo replied:
yank it :-)

IdanCo commented on src/library/pagination/pagination.component.js:
```diff
@@ -0,0 +1,170 @@
+
+    // validate bindings
+    if (typeof this.itemsPerPage !== 'number' || isFinite(this.itemsPerPage) === false) {
+      this.$log.error('invalid ngbsPagination::itemsPerPage: ', this.itemsPerPage, 'expecting a number');
```
LOVE the error handling and reporting! i'm going to implement it project wide

IdanCo commented on src/library/pagination/pagination.component.js:
```diff
@@ -0,0 +1,170 @@
+
+  buildPages() {
+    // determine the number of pages
+    this.totalPages = Math.ceil(this.totalItems / this.itemsPerPage);
```
I think we have a bug here:
Notice the first example in the demo. Event though you supplied 100 items, only 10 are available. This line is where it happens. Tried to debug it myself, i'm sure you'll do a much better job.

I think the underlying issue is terminology. It's not clear what's a page and what's an item. This is true for the md and the attributes as well. I suggest the following terminology -
**page** - currently referred to as _item_ or _page_ interchangeably. 
**batch** - currently referred to as _page_.

So with this new terminology, the attributes will change thus -
current-page --> no change
items-per-page --> pages-per-batch
total-items --> total-pages
visible-page-buffer -> no change

feel free to suggest a different terminology...

langdonx replied:
Ah, definitely a terminology difference.  I am totally open to any changes, but I tell you how I got here.

I had 10k photos to display on a screen.  That's too many for 1 screen, so I added pagination.  The original method signature to display the pagination (the series of numbers, representing groups of photos) looked like this (it's PHP, 🤢):

`print_page_index($p_total_pages, $p_current_page_number, $p_width = 4)`

This was great because (IMO) because it was so plainly obvious what it was looking.  If I have 100 photos... and I want to show 10 per screen.... I'd calculate that there should be 10 total pages.  Unfortunately because of the signature, I had to do that calculation every time I wanted to call `print_page_index` (for photos, forum posts, users, etc):

    ceil($total_number_of_photos_available / $total_number_of_photos_to_show_per_screen);

Years later I started using UI Bootstrap and have used it a ton.  Their attributes are identical to what I created here (with the exception of visible-page-buffer (because I took a different approach) and disabled (because I don't like to use `ng-` namespace for components I write).  I loved the fact that I didn't have to run that `ceil()` calculation anymore.  I loved that their *component signature* took arguments that reduced the amount of code I had to write.

I didn't intend to have **item** and **page** used interchangeably (perhaps I wrote some bad documentation).  A page to me (and to UI Bootstrap) is a group of items to be displayed together.  Much like pages of a book are simply groups of words -- where pagination is used to quickly reference a set of words that might be useful (or in the component's case  a set of *items*).  I don't really like the word "items", but I understand its generic to cover anything that the user might want to paginate.

Let me know if that doesn't clear up the attribute names or if you want to discuss further in a better medium (skype/hangouts/slack/discord/email/etc).

IdanCo replied:
ahh! ok! i think i actually heard the penny drops...

So terminology is great as it is, and the so-called bug is actually a feature (lol). If it's alright with you, i'll perhaps rephrase some of the MD in case there are more slow-minded folks like me out there. Don't worry about it now though.

IdanCo replied:
> (skype/hangouts/slack/discord/email/etc).

you forgot gitter - https://gitter.im/ng1bs4

IdanCo commented on src/library/pagination/pagination.component.js:
```diff
@@ -0,0 +1,170 @@
+      this.$log.error('invalid ngbsPagination::totalItems: ', this.totalItems, 'expecting a number');
+    }
+
+    this.$scope.$watch(() => `${this.currentPage}${this.itemsPerPage}${this.totalItems}${this.visiblePageBuffer}`, () => {
```
why no take advantage of the wonderful [lifecycle hooks components](https://docs.angularjs.org/guide/component#component-based-application-architecture) come with? specifically $onChanges(changesObj). Not obligatory, just a suggestion...

langdonx replied:
Total ignorance.  `$onChanges` looks as life changing as `angular.components`!  I'll refactor.

IdanCo commented on src/library/pagination/pagination.scss:
```diff
@@ -0,0 +1,16 @@
+    // it would be helpful to have access to user's variables here... i'm not sure if that's possible though.
+    // it's not great that the user can't see the selected page when ngbs-pagination is disabled
+    // it's also not great to hard-code these color values
+    // perhaps we leave it as is, and the user can override the style if they choose
```
I totally agree!!!

It's a real shame not to rely on the abundance of variable bs4 comes with. But to allow this, the build process will have to export a single sass file that can be imported by the library-user into their own sass build. This requires some redesigning...

IdanCo reviewed (commented):
Beautiful code, great addition to the library!

langdonx committed:
yanked redefinition of bindings

langdonx committed:
implemented $onChanges, added extra validation

langdonx committed:
removed double space

langdonx committed:
improved documentation

langdonx committed:
improved demo

langdonx commented:
Resolved the issues we discussed above:
 - Removed unnecessary constructor assignments.
 - Switched to `$onChanges`.  I was surprised to see that it fires before `$onInit`, so I had to move defaults to `$onChanges`, but left validation errors in `$onInit` so the console isn't flooded with them.
 - Improved documentation to hopefully make things more clear.  I tried to convey items is the number of things from *their own dataset*.
 - Improved the demo by reducing the number of pagination components displayed and adding inputs to control rendering.

Let me know if I missed anything, you catch any other weirdness, or you have any other ideas.  We could probably add more options, but it seems like a good start.

IdanCo commented:
Thank you so much for the awesome work!

IdanCo commented:
btw - $onChanges receives a changeObject with a flag indicating the controller is initializing. But it can easily wait for the next PR...
